### PR TITLE
nixos test ref/02 normalise test driver nix expressions

### PR DESCRIPTION
**Pull request by blaggacao**
- nixos/tests/test-driver: cleanup nix expression

---

Less nesting, where that improves readability. More nesteing, where
that improves readability, but most importantly:

Expose individual functions separately so that they can be more easily
built directly, eg.:

`nix build --impure --expr '(import ./testing-python.nix {system = builtins.currentSystem;}).mkTestDriver'`

Things done:

- factor `mkDriver` out of `makeTest` to clarify the different API-Levels (`makeTest` is a wrapper for test consumption that implements a attrset that satisfies external contracts, while `mkDriver` is a "convenience" script wrapper for `mkTestDriver`)
- fix the script name to `"nixos-test-driver"` (cosmetics to avoid it is reported as `.nixos-test-driver-wrapped`)
- `driver-exports` -> `driver-symbols` (that's what they are)
- simplify a bit conditional logic via "defaults over conditionals" (regarding `virtualisation.qemu.package`)

---
(this is an addition to https://github.com/NixOS/nixpkgs/pull/125971, which is merged)

### CODE REVIEW

**Comment by blaggacao** [comment]
@tfc Now this diff is review friendly. Thanks for merging #125971.

File: nixos/lib/testing-python.nix
Comment (blaggacao) [inline nixos/lib/testing-python.nix:68]: It doesn't nothing, but it would be the correct way to do it. I think for this to work throughout the whole nix ecosystem, python would need patching to effectively receive this arg0 from `exec -a`.

File: nixos/lib/test-driver/test-driver.py
Comment (tfc) [inline nixos/lib/test-driver/test-driver.py:976]: i stumble on this one: it appears evitable to do raw text-sustitution-preprocessing on the test script. if we start doing this, then we could/should substitute a lot of other things, too.

Comment (tfc) [reply nixos/lib/test-driver/test-driver.py:976]: ok it's too early for me, it is about substitution of course.

but i don't understand why we need to mix such things again. initially, you wanted to refactor the driver a bit to prepare for the upcoming tmux changes. now i am reviewing all kinds of different things that i wasn't expecting and this with all the coupled discussions drags too much time. i won't be able to timely accompany such efforts if we keep mixing things up.

Comment (blaggacao) [reply nixos/lib/test-driver/test-driver.py:976]: No worries! :wink:

This substitution, as the comment in `testing-python.nix` suggests is because 
`exec -a` doesn't seem to have effect for python shebangs.

So if you run the script and get the help text, then it would show as `.nixos-test-driver-wrapper` which does not correspond to what you did and hence is confusing ("a lie").

File: nixos/lib/testing-python.nix
Comment (blaggacao) [reply nixos/lib/testing-python.nix:68]: @tfc this is the rational with regard to your sister comment.

**Comment by tfc** [comment]
i don't want to review this. you opened multiple pull requests that go much further than the initial "pre tmux work refactoring", fix multiple things at once that were not discussed before and i have no idea why this all.

**Comment by blaggacao** [comment]
> i don't want to review this.

No problem. As a code owner, your review is automatically requested by github.

For the series of PR that are in my pipeline, you might probably just want to ignore those automatic requests.

I'll try to coordinate with the other folks that where involved in this motion that I'm (admittedly: creatively) spearheading here.

For the rest, let's take it back to matrix anytime you wish. :+1:

**Comment by tfc** [comment]
as a code owner i am also interested in knowing why certain changes are made. all the changes that are made to the nixos integration test driver are usually well explained, but not in this case. you're "fixing" random things and mix that with the changes that were initially supposed to be made, without explaining it.

i completely oppose that we do text substitution on the name of the binary, that everything is changed from env vars to parameters (but still supports env vars), etc. etc., this all looks like random rewrites  for no reason.

**Comment by blaggacao** [comment]
@tfc I'm confused. Does it mean you will be actually blocking this PR? Do you want to have a chat on matrix to expand on context and understanding? Doe you have difficulty to see the big picture as prototyped in the original PR?

**Comment by tfc** [comment]
i did not agree on all parts of the original PR, and all you did was explaining me why my reasoning is wrong or extreme. then you talked about "deals" and then opened a lot of PRs that intertwine a lot of additional changes that make no sense to me, regarding the initial discussion. what ought to be a refactoring that makes it easier to get a certain feature in completely blew up in a random rewrite of half the test driver, for no good technical reason.

i want to achieve that the integration test driver stays free of regression and only gets changes that either solve actual problems or improve the clarity of the architecture. and i don't see these goals matched here.

i am busy, and i imagine that you are, too, and i don't see how this style of working helps getting important things done with as few overhead as possible.

**Comment by blaggacao** [comment]
> and i don't see these goals matched here.

Well, in addition to not seeing [...], let me put on the record you are _also_ not available for direct conversation to clarify things and you _also_ seem not to consider arguments that are brought to your knowledge. (Like for example the motivation that is held by the commit message for that particular change-set in this very PR. But I can remember other instances.)

> i don't see how this style of working helps getting important things done with as few overhead as possible.

On behalf of the above I completely agree and deem necessary a community mediation, if such is available.

File: nixos/lib/testing-python.nix
Comment (tfc) [inline nixos/lib/testing-python.nix:19]:
```suggestion
  # Reifies and correctly wraps the python test driver for
```

File: nixos/lib/testing-python.nix
Comment (tfc) [inline nixos/lib/testing-python.nix:99]: `mkTestDriver` and `mkDriver` are somehow not the most self-describing names.

i think the `mkTestDriver` attribute could just be `pythonTestDriver` or so (because it's just the naked python script wrapped with some generic runtime env info).
`mkDriver` could be `setupDriverForTest` or something like that (because we're taking the script and basically add all the needed parameters that are finally test-specific).

File: nixos/lib/testing-python.nix
Comment (tfc) [inline nixos/lib/testing-python.nix:133]: i was a bit irritated why the outer parentheses are here because it's not done like this in other lines of code

File: nixos/lib/testing-python.nix
Comment (tfc) [inline nixos/lib/testing-python.nix:169]: i found the old name `nodeHostNames` more self-explanatory, because it is really a list of host names.

File: nixos/lib/testing-python.nix
Comment (tfc) [inline nixos/lib/testing-python.nix:230]: in both invocations of `mkDriver` we let `build-vms.nix` generate VMs based on different qemu variations, but in both cases `mkDriver` is called with the same default argument `qemu_pkg ? pkgs.qemu_test`. We should make sure that we don't mix up qemu packages

**Review by tfc — CHANGES_REQUESTED** [review]

File: nixos/lib/testing-python.nix
Comment (blaggacao) [reply nixos/lib/testing-python.nix:169]: It might change in subsequent commits, but here that's orthogonal, you're right. I'm reverting.

File: nixos/lib/testing-python.nix
Comment (blaggacao) [reply nixos/lib/testing-python.nix:230]: Nice catch! Thx!

**Comment by blaggacao** [comment]
@GrahamcOfBorg test.basic

**Comment by blaggacao** [comment]
@GrahamcOfBorg test basic

**Comment by blaggacao** [comment]
@GrahamcOfBorg test boot

**Comment by blaggacao** [comment]
@GrahamcOfBorg test installer

**Comment by blaggacao** [comment]
@GrahamcOfBorg test bittorrent

**Comment by blaggacao** [comment]
@GrahamcOfBorg test installer.bcache installer.btrfsSimple installer.btrfsSubvolDefault installer.btrfsSubvols installer.encryptedFSWithKeyfile installer.grub1 installer.luksroot installer.lvm installer.separateBoot installer.simple installer.swraid installer.fsroot

**Comment by blaggacao** [comment]
@GrahamcOfBorg test boot.biosCdrom boot.Netboot boot.biosUsb boot.uefiCdrom boot.uefiNetboot boot.Usb

**Comment by blaggacao** [comment]
@GrahamcOfBorg test bittorrent

**Comment by blaggacao** [comment]
@GrahamcOfBorg test boot.biosCdrom boot.biosNetboot boot.biosUsb boot.uefiCdrom boot.uefiNetboot boot.uefiUsb

**Review by tfc — APPROVED** [review]

**Pull request merged**
